internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	requestTimeout    = 10 * time.Second
+	maxRequestEntries = 100
+)
+
+type Client struct {
+	ClientArgs
+
+	targetName string
+	remoteURL  string
+	entries    []lokiEntryWithService
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return NewClientWithArgs(target, ClientArgs{})
+}
+
+// ClientArgs allows overriding default parameters (e.g. for testing)
+type ClientArgs struct {
+	RequestTimeout    time.Duration
+	MaxRequestEntries int
+}
+
+func NewClientWithArgs(target *plan.LogTarget, args ClientArgs) *Client {
```

Comment:
These option values are typically passed as a pointer. As a general rule, unless the value semantics are intentional (think `time.Time`, for instance), it's typically passed by identity. Also, our naming convention for these cases is`FooOptions`. For some examples, there are several instances of both of these patterns in our `client` Pebble package.